events/src/utils.py

```diff
@@ -313,6 +314,40 @@ def plugin_webhook(message):
     webhooks_pool.send(message["event"], message["ts"], message_json)
 
 
+class NamedTempOpen(object):
+    """
+        This class is used to create a temporary file which is then written to with contents.
+        The temp file is then persisted with the give name by calling os.rename().
+        This class is used to avoid the data loss or truncation in case of multiple processes
+        writing to the same file without the use of fcntl locks.
+
+        The temporary file is created in the cwd of the process.
+    """
+
+    def __init__(self, filename, open_mode, *args, **kwagrs) -> None:
+        self.filename = filename
+        self.open_mode = open_mode
+        self.open_args = args
+        self.open_kwargs = kwagrs
+        self.fileobj = None
+
+    def __enter__(self):
+        tfile = NamedTemporaryFile(mode=self.open_mode,
+                                   delete=False,
+                                   dir=".",
+                                   *self.open_args,
+                                   **self.open_kwargs)
+
+        if tfile is None:
+            raise
+
+        self.fileobj = tfile
+        return self.fileobj
+
+    def __exit__(self, ex_type, ex_val, ex_tb):
+        self.fileobj.close()
+        os.rename(self.fileobj.name, self.filename)
```

Comment:
It doesn't seem to be specified if the name of the file will be present after `close()`. Probably it would be safer to read the name before closing or even do the rename before.

Also, if the exit has happened as the result of an exception, we shouldn't rename the file, just delete it.

Reply:
If we pass `delete=False` as a parameter to `NameTemporaryFile` it won't delete the file after `close()` so we manually have to take care of that. So in this case, doing the rename after `close` is fine.

Reply:
> Also, if the exit has happened as the result of an exception, we shouldn't rename the file, just delete it.

The `__exit__` block would not be called in the case of an exception. And the exception would only happen in case temp file is not created for some reason. So, no manual work needed there as the file would not exist.

Reply:
> If we pass `delete=False` as a parameter to `NameTemporaryFile` it won't delete the file after `close()` so we manually have to take care of that. So in this case, doing the rename after `close` is fine as the file will be present even after close (as long as we are not using os' temp dirs, which we are not).

I'm not saying that the file won't be there. What I'm saying is that the `name` field of the object may be undefined after closing the file. I haven't seen any clear documentation about that, but one example in python documentation is using the name after closing the file, so I think it should be ok.

> The `__exit__` block would not be called in the case of an exception. And the exception would only happen in case temp file is not created for some reason. So, no manual work needed there as the file would not exist.

`__exit__` won't be called in case of exception inside the `__enter__` method, but it will be called if something fails inside `with` block. An exception there means that the file may be corrupt or incomplete. Renaming it to the "good" name and making it visible to everyone doesn't seem a good idea.

Reply:
> Renaming it to the "good" name and making it visible to everyone doesn't seem a good idea.

Agreed. Leaving the temp file as is was also not a good idea, addressed that too. If an exception occurs inside `with`, the temp file is removed from the disk.

Reply:
Side note: Thank you for scrutinizing the patches. Got so many new things to learn :)